### PR TITLE
Preload precomputed release

### PR DESCRIPTION
**Pull request by se7entyse7en**
Closes [DEV-2148](https://athenianco.atlassian.net/browse/DEV-2148).

Tested with the following:
```
hey \
  -H 'Connection: keep-alive' \
  -H 'Pragma: no-cache' \
  -H 'Cache-Control: no-cache' \
  -H 'sec-ch-ua: "Google Chrome";v="89", "Chromium";v="89", ";Not A Brand";v="99"' \
  -H 'Accept: application/json' \
  -H 'Authorization: Bearer <evicted>' \
  -H 'sec-ch-ua-mobile: ?0' \
  -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 11_2_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36' \
  -H 'Content-Type: application/json' \
  -H 'Origin: http://localhost:3000' \
  -H 'Sec-Fetch-Site: same-site' \
  -H 'Sec-Fetch-Mode: cors' \
  -H 'Sec-Fetch-Dest: empty' \
  -H 'Referer: http://localhost:3000/' \
  -H 'Accept-Language: en-US,en;q=0.9,it-IT;q=0.8,it;q=0.7' \
  -m POST \
  -d '{"for":[{"repositories":["github.com/athenianco/athenian-api","github.com/athenianco/athenian-owl","github.com/athenianco/athenian-push-api","github.com/athenianco/athenian-webapp","github.com/athenianco/automation","github.com/athenianco/browser-extensions","github.com/athenianco/ci","github.com/athenianco/cloud-common","github.com/athenianco/environments","github.com/athenianco/helm-repo","github.com/athenianco/infrastructure","github.com/athenianco/metadata","github.com/athenianco/names-matcher","github.com/athenianco/pgbouncer-rr-patch","github.com/athenianco/precomputer","github.com/athenianco/reposet-sync"],"repogroups":[[0],[1],[2],[3],[4],[5],[6],[7],[8],[9],[10],[11],[12],[13],[14],[15]],"with":{"author":["github.com/vmarkovtsev","github.com/apps/dependabot","github.com/apps/dependabot-preview","github.com/gkwillie","github.com/snyk-bot","github.com/akbarik","github.com/dennwc","github.com/dpordomingo","github.com/lwsanty","github.com/se7entyse7en","github.com/znegrin","github.com/eiso","github.com/ibot3","github.com/marnovo","github.com/rrickay","github.com/warenlg"]}}],"metrics":["pr-wip-time","pr-wip-count","pr-wip-count-q","pr-review-time","pr-review-count","pr-review-count-q","pr-merging-time","pr-merging-count","pr-merging-count-q","pr-release-time","pr-release-count","pr-release-count-q","pr-lead-time","pr-lead-count","pr-lead-count-q","pr-cycle-time","pr-cycle-count","pr-cycle-count-q","pr-all-count","pr-wait-first-review-time","pr-wait-first-review-count","pr-wait-first-review-count-q","pr-flow-ratio","pr-opened","pr-reviewed","pr-not-reviewed","pr-merged","pr-rejected","pr-closed","pr-done","pr-size","pr-median-size","pr-wip-pending-count","pr-review-pending-count","pr-merging-pending-count","pr-release-pending-count","pr-opened-mapped-to-jira","pr-done-mapped-to-jira","pr-all-mapped-to-jira","pr-participants-per","pr-review-comments-per","pr-reviews-per"],"date_from":"2020-05-06","date_to":"2021-05-06","granularities":["all"],"exclude_inactive":true,"account":1,"timezone":120,"quantiles":[0,0.95]}' \
  -n 10 -c 1 -t 0 'http://localhost:8080/v1/metrics/prs'
```

# With preloading

```
Summary:
  Total:	11.5875 secs
  Slowest:	1.8847 secs
  Fastest:	0.9786 secs
  Average:	1.1587 secs
  Requests/sec:	0.8630


Response time histogram:
  0.979 [1]	|■■■■■■■■■■
  1.069 [4]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.160 [2]	|■■■■■■■■■■■■■■■■■■■■
  1.250 [2]	|■■■■■■■■■■■■■■■■■■■■
  1.341 [0]	|
  1.432 [0]	|
  1.522 [0]	|
  1.613 [0]	|
  1.703 [0]	|
  1.794 [0]	|
  1.885 [1]	|■■■■■■■■■■


Latency distribution:
  10% in 0.9919 secs
  25% in 1.0550 secs
  50% in 1.0788 secs
  75% in 1.2130 secs
  90% in 1.8847 secs
  0% in 0.0000 secs
  0% in 0.0000 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0005 secs, 0.9786 secs, 1.8847 secs
  DNS-lookup:	0.0002 secs, 0.0000 secs, 0.0016 secs
  req write:	0.0001 secs, 0.0000 secs, 0.0010 secs
  resp wait:	1.1577 secs, 0.9783 secs, 1.8786 secs
  resp read:	0.0003 secs, 0.0002 secs, 0.0009 secs

Status code distribution:
  [200]	10 responses
```

# No preloading

```
Summary:
  Total:	14.9071 secs
  Slowest:	2.1123 secs
  Fastest:	1.0840 secs
  Average:	1.4907 secs
  Requests/sec:	0.6708


Response time histogram:
  1.084 [1]	|■■■■■■■■■■■■■■■■■■■■
  1.187 [1]	|■■■■■■■■■■■■■■■■■■■■
  1.290 [1]	|■■■■■■■■■■■■■■■■■■■■
  1.392 [2]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.495 [0]	|
  1.598 [2]	|■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■■
  1.701 [0]	|
  1.804 [1]	|■■■■■■■■■■■■■■■■■■■■
  1.907 [1]	|■■■■■■■■■■■■■■■■■■■■
  2.009 [0]	|
  2.112 [1]	|■■■■■■■■■■■■■■■■■■■■


Latency distribution:
  10% in 1.1699 secs
  25% in 1.3129 secs
  50% in 1.5033 secs
  75% in 1.8290 secs
  90% in 2.1123 secs
  0% in 0.0000 secs
  0% in 0.0000 secs

Details (average, fastest, slowest):
  DNS+dialup:	0.0002 secs, 1.0840 secs, 2.1123 secs
  DNS-lookup:	0.0001 secs, 0.0000 secs, 0.0009 secs
  req write:	0.0001 secs, 0.0000 secs, 0.0002 secs
  resp wait:	1.4901 secs, 1.0838 secs, 2.1112 secs
  resp read:	0.0003 secs, 0.0002 secs, 0.0010 secs

Status code distribution:
  [200]	10 responses
```

# Speedup

So far two queries have been ported and the speedup is around 1.5x. Unfortunately, I tested locally with only our own data as I'm having issues with resources for restoring the production db, I'll try it again in the next days.It's likely that the speedup for bigger accounts is already much higher than 1.5x considering that when PG plans get mad the performance can be really bad. In any case, testing with bigger accounts should answer this.

Screenshot 2021-05-06 at 19 12 03

As can be seen from the image, the `ReleaseLoader._fetch_precomputed_releases` got a speedup of ~16x, and `PullRequestMiner._fetch_prs_by_filters` got a speedup of ~3x. To be noted that these values are just from a single observation that is the image above, and not coming from a distribution as done for the endpoint wall time. But it's useful to have a rough idea as well of the low-level gain.

# Memory usage for preloading

```
# DB: mdb                                          | [total: 25.66 MiB]

## DataFrame: prs                                  | [total: 25.66 MiB]

# DB: pdb                                          | [total: 904.98 KiB]

## DataFrame: releases                             | [total: 904.98 KiB]
```

### CODE REVIEW

**Comment by se7entyse7en** [comment]
@vmarkovtsev I'm curious to hear if you have some further thoughts/ideas.

**Comment by se7entyse7en** [comment]
I'm also curious how it would improve once the pandas optimizations you proposed will be included as well.

**Comment by se7entyse7en** [comment]
For sure even preloading only specific accounts will help as it will need to filter smaller dataframes.

**Comment by vmarkovtsev** [comment]
> I'm also curious how it would improve once the pandas optimizations you proposed will be included as well.
Regarding this, once we push this to staging we'll see crystal clear traces in Sentry that will reveal the new bottlenecks. Eager to study them :nerd_face:

File: server/athenian/api/controllers/features/entries.py [hunk 1]
```diff
@@ -77,6 +77,7 @@ class MetricEntriesCalculator:
     """Calculator for different metrics."""
 
     miner = PullRequestMiner
+    unfresh_fetcher = UnfreshPullRequestFactsFetcher
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/features/entries.py:80]:
```suggestion
    unfresh_pr_facts_fetcher = UnfreshPullRequestFactsFetcher
```

File: server/athenian/api/controllers/miners/github/release_load.py [hunk 2]
```diff
@@ -665,11 +684,16 @@ def match_groups_to_sql(match_groups: Dict[ReleaseMatch, Dict[str, Iterable[str]
...
-        or_items.extend(and_(model.release_match == "".join([match.name, suffix, v]),
-                             model.repository_full_name.in_(r))
-                        for v, r in match_group.items())
+
+        or_conditions.extend([
+            {
+                "release_match": "".join([match.name, suffix, v]),
+                "repository_full_name": r,
+            } for v, r in match_group.items()
+        ])
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/miners/github/release_load.py:693]:
```suggestion
        or_conditions.extend({
            "release_match": "".join([match.name, suffix, v]),
            "repository_full_name": r,
        } for v, r in match_group.items())
```
(more compact formatting + no need to allocate a list to extend)

File: server/athenian/api/experiments/preloading/entries.py [hunk 3]
```diff
@@ -1,27 +1,93 @@
...
+        or_masks = [
+            (df["release_match"] == cond["release_match"]) &
+            (df["repository_full_name"].isin(cond["repository_full_name"]))
+            for cond in or_conditions
+        ]
+
+        return functools.reduce(operator.or_, or_masks)
+
+
+class ReleaseToPullRequestMapper(OriginalReleaseToPullRequestMapper):
```
Comment (vmarkovtsev) [inline server/athenian/api/experiments/preloading/entries.py:70]: Maybe not in this PR, but I'd rather see this naming scheme:

Original -> (delete)
New classes -> prepend Preloaded

Thus

```suggestion
class PreloadedReleaseToPullRequestMapper(ReleaseToPullRequestMapper):
```

File: server/athenian/api/experiments/preloading/entries.py [hunk 4]
```diff
@@ -1,27 +1,93 @@
...
+    """Mapper from releases to pull requests."""
+
+    release_loader = ReleaseLoader
+
+
+class UnfreshPullRequestFactsFetcher(OriginalUnfreshPullRequestFactsFetcher):
+    """Fetcher for unfreshed pull requests facts."""
```
Comment (vmarkovtsev) [inline server/athenian/api/experiments/preloading/entries.py:77]:
```suggestion
    """Fetcher for unfresh pull requests facts."""
```
BTW, with or without preloading?

File: server/athenian/api/experiments/preloading/entries.py [hunk 5]
```diff
@@ -1,27 +1,93 @@
...
+from athenian.api.controllers.miners.github.release_match import \
+    ReleaseToPullRequestMapper as OriginalReleaseToPullRequestMapper
 from athenian.api.controllers.miners.types import PRParticipants, PRParticipationKind
+from athenian.api.controllers.settings import ReleaseMatch
 from athenian.api.models.metadata.github import PullRequest
 from athenian.api.tracing import sentry_span
 
 
+class ReleaseLoader(OriginalReleaseLoader):
+    """Loader for releases."""
```
Comment (vmarkovtsev) [inline server/athenian/api/experiments/preloading/entries.py:31]: with or without preloading?

File: server/athenian/api/controllers/miners/github/release_load.py [hunk 6]
```diff
@@ -654,8 +654,27 @@ def match_groups_to_sql(match_groups: Dict[ReleaseMatch, Dict[str, Iterable[str]
...
+    return or_items, repos
+
+
+def match_groups_to_conditions(
+    match_groups: Dict[ReleaseMatch, Dict[str, Iterable[str]]],
+) -> Tuple[List[List[dict]], List[Iterable[str]]]:
+    """
+    Convert the grouped release matches to a list of conditions.
+
+    :return: 1. List of the alternative conditions. \
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/miners/github/release_load.py:674]:
```suggestion
    :return: 1. List of the filters to OR/UNION later. \
```

File: server/athenian/api/controllers/miners/github/release_load.py [hunk 7]
```diff
@@ -654,8 +654,27 @@ def match_groups_to_sql(match_groups: Dict[ReleaseMatch, Dict[str, Iterable[str]
...
+             2. List of involved repository names for each SQL filter.
```
Comment (vmarkovtsev) [inline server/athenian/api/controllers/miners/github/release_load.py:675]:
```suggestion
             2. List of involved repository names for each filter.
```

**Review by vmarkovtsev — CHANGES_REQUESTED** [review]

**Pull request merged**